File: Sources/RedlineCore/Trends.swift
// Time-bucketed usage for the dashboard charts. Every entry already carries a timestamp, so
// history comes from the transcripts themselves; nothing extra is recorded.
import Foundation

public struct UsagePoint: Equatable {
    public let start: Date
    public let io: Int
    public let cost: Double
    public let cacheRead: Int
    public let cacheWrite: Int

    public init(start: Date, io: Int, cost: Double, cacheRead: Int, cacheWrite: Int) {
        self.start = start
        self.io = io
        self.cost = cost
        self.cacheRead = cacheRead
        self.cacheWrite = cacheWrite
    }
}

public struct ProviderTrend: Equatable {
    public let provider: String
    public let points: [UsagePoint]

    public var totalIO: Int { points.reduce(0) { $0 + $1.io } }
    public var totalCost: Double { points.reduce(0) { $0 + $1.cost } }
    public var peak: UsagePoint? { points.max(by: { $0.io < $1.io }) }
}

public struct ModelShare: Equatable {
    public let model: String
    public let provider: String
    public let io: Int
    public let cost: Double
    public let priced: Bool
}

public enum Bucketing: Equatable {
    case day
    case hour

    var component: Calendar.Component { self == .day ? .day : .hour }
}

/// How far apart the dated labels on a daily axis should sit.
///
/// Lives here rather than in the chart so it can be tested: the whole point is that a quarter
/// does not print a label every five days and turn the axis into a smear, and that both daily
/// charts use one cadence so the same range cannot read as two different spans stacked up.
public enum DailyAxis {
    /// Roughly six to eight labels at any range RedLine offers.
    public static func strideDays(for range: Int) -> Int {
        switch range {
        case ...7:   return 1
        case ...14:  return 2
        case ...30:  return 5
        case ...60:  return 10
        default:     return 14
        }
    }

    static let months = ["Jan", "Feb", "Mar", "Apr", "May", "Jun",
                         "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]

    /// The day a bucket belongs to, unambiguously. Built by hand rather than with a
    /// DateFormatter, which is where the cost formatting went wrong on Linux.
    public static func key(for start: Date, calendar: Calendar = .current) -> String {
        let parts = calendar.dateComponents([.year, .month, .day], from: start)
        return String(format: "%04d-%02d-%02d",
                      parts.year ?? 0, parts.month ?? 0, parts.day ?? 0)
    }

    /// What an axis prints. Published by the engine so two shells cannot label one day two
    /// ways, and English on purpose, like every other string RedLine shows.
    public static func label(for start: Date, calendar: Calendar = .current) -> String {
        let parts = calendar.dateComponents([.month, .day], from: start)
        let month = min(max((parts.month ?? 1) - 1, 0), 11)
        return "\(months[month]) \(parts.day ?? 0)"
    }
}

public enum Trends {
    // Buckets are pre-filled with zeros across the whole range so a chart shows a quiet day
    // as a gap at the baseline rather than skipping the date entirely.
    public static func trend(_ entries: [Entry],
                             by bucketing: Bucketing,
                             count: Int,
                             now: Date = Date(),
                             calendar: Calendar = .current,
                             config: Config) -> [ProviderTrend] {
        guard count > 0 else { return [] }
        let starts = bucketStarts(by: bucketing, count: count, now: now, calendar: calendar)
        guard let first = starts.first else { return [] }

        // provider -> bucket start -> running totals
        var acc: [String: [Date: (io: Int, cost: Double, cr: Int, cw: Int)]] = [:]
        for e in entries where e.ts >= first {
            guard let start = calendar.dateInterval(of: bucketing.component, for: e.ts)?.start
            else { continue }
            let cost = self.cost(of: e, config: config)
            var slot = acc[e.provider]?[start] ?? (0, 0, 0, 0)
            slot.io += e.input + e.output
            slot.cost += cost
            slot.cr += e.cacheRead
            slot.cw += e.cache5m + e.cache1h
            acc[e.provider, default: [:]][start] = slot
        }

        return acc.keys.sorted().map { provider in
            let byStart = acc[provider] ?? [:]
            let points = starts.map { start -> UsagePoint in
                let s = byStart[start] ?? (0, 0, 0, 0)
                return UsagePoint(start: start, io: s.io, cost: s.cost,
                                  cacheRead: s.cr, cacheWrite: s.cw)
            }
            return ProviderTrend(provider: provider, points: points)
        }
    }

    static func bucketStarts(by bucketing: Bucketing, count: Int,
                             now: Date, calendar: Calendar) -> [Date] {
        guard let current = calendar.dateInterval(of: bucketing.component, for: now)?.start
        else { return [] }
        // Oldest first, ending with the bucket now falls in
        return (0..<count).reversed().compactMap {
            calendar.date(byAdding: bucketing.component, value: -$0, to: current)
        }
    }

    /// Everything a chart in another language needs, in one shape.
    ///
    /// Pure on purpose: the command gathers the data, this decides what is published, and a
    /// test can pin every value without a warehouse or a clock. See SettingsContractTests'
    /// twin for trends, which both languages read.
    public static func report(days: Int, series: [UsagePoint], providers: [ProviderTrend],
                              models: [ModelShare],
                              calendar: Calendar = .current) -> [String: Any] {
        [
            "days": days,
            "label_every_days": DailyAxis.strideDays(for: days),
            "day_basis": "local",
            "tokens": models.reduce(0) { $0 + $1.io },
            "cost_usd": models.reduce(0.0) { $0 + $1.cost },
            // One unpriced model makes every total a floor, and saying so is the whole
            // difference between a number and a guess
            "has_unpriced": models.contains { !$0.priced },
            "series": series.map { point($0, calendar: calendar) },
            "providers": providers.map {
                ["provider": $0.provider, "tokens": $0.totalIO, "cost_usd": $0.totalCost,
                 "points": $0.points.map { point($0, calendar: calendar) }]
            },
            "models": models.map {
                ["model": $0.model, "provider": $0.provider, "tokens": $0.io,
                 "cost_usd": $0.cost, "priced": $0.priced]
            },
        ]
    }

    static func point(_ spot: UsagePoint, calendar: Calendar) -> [String: Any] {
        ["day": DailyAxis.key(for: spot.start, calendar: calendar),
         "label": DailyAxis.label(for: spot.start, calendar: calendar),
         "tokens": spot.io, "cost_usd": spot.cost]
    }

    /// Every provider's buckets added together, for a chart that draws one series. The
    /// buckets are the same starts in the same order for every provider, so this lines up.
    public static func combine(_ trends: [ProviderTrend]) -> [UsagePoint] {
        guard let first = trends.first else { return [] }
        return first.points.indices.map { i in
            var io = 0
            var cost = 0.0
            var cacheRead = 0
            var cacheWrite = 0
            for trend in trends where i < trend.points.count {
                io += trend.points[i].io
                cost += trend.points[i].cost
                cacheRead += trend.points[i].cacheRead
                cacheWrite += trend.points[i].cacheWrite
            }
            return UsagePoint(start: first.points[i].start, io: io, cost: cost,
                              cacheRead: cacheRead, cacheWrite: cacheWrite)
        }
    }

    public static func byModel(_ entries: [Entry], since: Date,
                               config: Config) -> [ModelShare] {
        var acc: [String: (provider: String, io: Int, cost: Double, priced: Bool)] = [:]
        for e in entries where e.ts >= since {
            let priced = config.price(for: e.model) != nil
            var slot = acc[e.model] ?? (e.provider, 0, 0, priced)
            slot.io += e.input + e.output
            slot.cost += cost(of: e, config: config)
            slot.priced = priced
            acc[e.model] = slot
        }
        return acc.map {
            ModelShare(model: $0.key, provider: $0.value.provider, io: $0.value.io,
                       cost: $0.value.cost, priced: $0.value.priced)
        }
        // Largest first, with a stable tiebreak so the order does not jitter between refreshes
        .sorted { $0.io == $1.io ? $0.model < $1.model : $0.io > $1.io }
    }

    static func cost(of e: Entry, config: Config) -> Double {
        guard let p = config.price(for: e.model) else { return 0 }
        var total = Double(e.input) * p.input
        total += Double(e.output) * p.output
        total += Double(e.cacheRead) * p.cacheRead
        total += Double(e.cache5m) * p.input * 1.25
        total += Double(e.cache1h) * p.input * 2.0
        return total / 1_000_000
    }
}

File: windows/src/Redline.Core/Trends.cs
// Time-bucketed usage for the dashboard charts. Every entry already carries a timestamp, so
// history comes from the transcripts themselves; nothing extra is recorded.
namespace Redline.Core;

public sealed record UsagePoint(DateTimeOffset Start, int Io, double Cost, int CacheRead, int CacheWrite);

public sealed record ProviderTrend(string Provider, IReadOnlyList<UsagePoint> Points)
{
    public int TotalIO => Points.Sum(p => p.Io);
    public double TotalCost => Points.Sum(p => p.Cost);
    public UsagePoint? Peak => Points.Count == 0 ? null : Points.MaxBy(p => p.Io);
}

public sealed record ModelShare(string Model, string Provider, int Io, double Cost, bool Priced);

public enum Bucketing { Day, Hour }

/// How far apart the dated labels on a daily axis sit. Tested here so a quarter does not
/// print a label every five days, and both daily charts share one cadence.
public static class DailyAxis
{
    /// Roughly six to eight labels at any range RedLine offers.
    public static int StrideDays(int range) => range switch
    {
        <= 7 => 1,
        <= 14 => 2,
        <= 30 => 5,
        <= 60 => 10,
        _ => 14,
    };
}

public static class Trends
{
    // Buckets are pre-filled with zeros across the whole range so a chart shows a quiet day
    // as a gap at the baseline rather than skipping the date entirely.
    public static List<ProviderTrend> Trend(IEnumerable<Entry> entries, Bucketing by, int count,
                                            Config config, DateTimeOffset? now = null,
                                            TimeZoneInfo? timeZone = null)
    {
        if (count <= 0) return new();
        var tz = timeZone ?? TimeZoneInfo.Local;
        var starts = BucketStarts(by, count, now ?? DateTimeOffset.UtcNow, tz);
        if (starts.Count == 0) return new();
        var first = starts[0];

        // provider -> bucket start -> running totals
        var acc = new Dictionary<string, Dictionary<DateTimeOffset, (int Io, double Cost, int Cr, int Cw)>>(StringComparer.Ordinal);
        foreach (var e in entries)
        {
            if (e.Ts < first) continue;
            var start = StartOf(by, e.Ts, tz);
            var cost = Cost(e, config);
            if (!acc.TryGetValue(e.Provider, out var byStart)) acc[e.Provider] = byStart = new();
            (int Io, double Cost, int Cr, int Cw) slot = byStart.TryGetValue(start, out var s) ? s : (0, 0.0, 0, 0);
            slot.Io += e.Input + e.Output;
            slot.Cost += cost;
            slot.Cr += e.CacheRead;
            slot.Cw += e.Cache5m + e.Cache1h;
            byStart[start] = slot;
        }

        return acc.Keys.OrderBy(k => k, StringComparer.Ordinal).Select(provider =>
        {
            var byStart = acc[provider];
            var points = starts.Select(start =>
            {
                var s = byStart.TryGetValue(start, out var x) ? x : (0, 0.0, 0, 0);
                return new UsagePoint(start, s.Item1, s.Item2, s.Item3, s.Item4);
            }).ToList();
            return new ProviderTrend(provider, points);
        }).ToList();
    }

    internal static List<DateTimeOffset> BucketStarts(Bucketing by, int count, DateTimeOffset now,
                                                      TimeZoneInfo timeZone)
    {
        var current = StartOf(by, now, timeZone);
        var output = new List<DateTimeOffset>(Math.Max(0, count));
        // Oldest first, ending with the bucket now falls in
        for (int back = count - 1; back >= 0; back--)
        {
            if (by == Bucketing.Hour)
            {
                output.Add(current.AddHours(-back));
            }
            else
            {
                // Calendar arithmetic on days keeps local midnight across a DST change
                var local = TimeZoneInfo.ConvertTime(current, timeZone).DateTime.Date.AddDays(-back);
                output.Add(FromLocal(local, timeZone));
            }
        }
        return output;
    }

    /// Start of the local day or hour `instant` falls in, as a UTC instant.
    public static DateTimeOffset StartOf(Bucketing by, DateTimeOffset instant, TimeZoneInfo timeZone)
    {
        var local = TimeZoneInfo.ConvertTime(instant, timeZone);
        var dt = local.DateTime;
        if (by == Bucketing.Hour)
        {
            // Keep the instant's own offset so the two hours of a fall-back night stay distinct
            var truncated = new DateTime(dt.Year, dt.Month, dt.Day, dt.Hour, 0, 0, DateTimeKind.Unspecified);
            var candidate = new DateTimeOffset(truncated, local.Offset);
            if (candidate <= instant && instant - candidate < TimeSpan.FromHours(1)) return candidate.ToUniversalTime();
            return FromLocal(truncated, timeZone);
        }
        return FromLocal(dt.Date, timeZone);
    }

    // A wall-clock time to an instant: a skipped time moves forward to the first valid one,
    // an ambiguous one takes the earlier instant, as Foundation's Calendar does
    private static DateTimeOffset FromLocal(DateTime local, TimeZoneInfo timeZone)
    {
        var t = DateTime.SpecifyKind(local, DateTimeKind.Unspecified);
        int guard = 0;
        while (timeZone.IsInvalidTime(t) && guard++ < 24 * 60) t = t.AddMinutes(1);
        TimeSpan offset = timeZone.IsAmbiguousTime(t)
            ? timeZone.GetAmbiguousTimeOffsets(t).Max()
            : timeZone.GetUtcOffset(t);
        return new DateTimeOffset(t, offset).ToUniversalTime();
    }

    public static List<ModelShare> ByModel(IEnumerable<Entry> entries, DateTimeOffset since, Config config)
    {
        var acc = new Dictionary<string, (string Provider, int Io, double Cost, bool Priced)>(StringComparer.Ordinal);
        foreach (var e in entries)
        {
            if (e.Ts < since) continue;
            var priced = config.Price(e.Model) is not null;
            (string Provider, int Io, double Cost, bool Priced) slot = acc.TryGetValue(e.Model, out var s) ? s : (e.Provider, 0, 0.0, priced);
            slot.Io += e.Input + e.Output;
            slot.Cost += Cost(e, config);
            slot.Priced = priced;
            acc[e.Model] = slot;
        }
        // Largest first, with a stable tiebreak so the order does not jitter between refreshes
        return acc.Select(kv => new ModelShare(kv.Key, kv.Value.Provider, kv.Value.Io, kv.Value.Cost, kv.Value.Priced))
                  .OrderByDescending(m => m.Io).ThenBy(m => m.Model, StringComparer.Ordinal).ToList();
    }

    internal static double Cost(Entry e, Config config) => Usage.CostOf(e, config) ?? 0;
}
